pkg/kubelet/kubeletconfig/checkpoint/download.go

```diff
@@ -138,32 +167,91 @@ func (r *remoteConfigMap) APIPath() string {
+		cm, err = client.CoreV1().ConfigMaps(r.source.ConfigMap.Namespace).Get(r.source.ConfigMap.Name, metav1.GetOptions{})
+		if err != nil {
+			return nil, status.DownloadError, fmt.Errorf("%s, error: %v", status.DownloadError, err)
+		}
+		utillog.Infof("successfully downloaded %s, UID: %s, ResourceVersion: %s", r.APIPath(), cm.UID, cm.ResourceVersion)
+	} // Assert: Now we have a non-nil ConfigMap
+	// construct Payload from the ConfigMap
+	payload, err := NewConfigMapPayload(cm)
 	if err != nil {
-		return nil, status.DownloadError, fmt.Errorf("%s, error: %v", status.DownloadError, err)
+		// We only expect an error here if ObjectMeta is lacking UID or ResourceVersion. This should
+		// never happen on objects in the informer's store, or objects downloaded from the API server
+		// directly, so we report InternalError.
+		return nil, status.InternalError, fmt.Errorf("%s, error: %v", status.InternalError, err)
 	}
+	// update internal UID and ResourceVersion based on latest ConfigMap
+	r.source.ConfigMap.UID = cm.UID
+	r.source.ConfigMap.ResourceVersion = cm.ResourceVersion
+	return payload, "", nil
+}
 
-	// ensure that UID matches the UID on the source
-	if r.source.ConfigMap.UID != cm.UID {
-		reason = fmt.Sprintf(status.UIDMismatchErrorFmt, r.source.ConfigMap.UID, r.APIPath(), cm.UID)
-		return nil, reason, fmt.Errorf(reason)
+func (r *remoteConfigMap) Informer(client clientset.Interface,
+	addFunc func(newObj interface{}),
+	updateFunc func(oldObj interface{}, newObj interface{}),
+	deleteFunc func(deletedObj interface{})) cache.SharedInformer {
+	// select ConfigMap by name
+	fieldselector := fields.OneTermEqualSelector("metadata.name", r.source.ConfigMap.Name)
+
+	// add some randomness to resync period, which can help avoid controllers falling into lock-step
+	minResyncPeriod := 15 * time.Minute
+	factor := rand.Float64() + 1
+	resyncPeriod := time.Duration(float64(minResyncPeriod.Nanoseconds()) * factor)
+
+	lw := &cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (kuberuntime.Object, error) {
+			return client.CoreV1().ConfigMaps(r.source.ConfigMap.Namespace).List(metav1.ListOptions{
+				FieldSelector: fieldselector.String(),
+			})
+		},
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			return client.CoreV1().ConfigMaps(r.source.ConfigMap.Namespace).Watch(metav1.ListOptions{
+				FieldSelector:   fieldselector.String(),
+				ResourceVersion: options.ResourceVersion,
+			})
+		},
 	}
 
-	payload, err := NewConfigMapPayload(cm)
-	if err != nil {
-		reason = fmt.Sprintf("invalid downloaded object")
-		return nil, reason, fmt.Errorf("%s, error: %v", reason, err)
+	handler := cache.ResourceEventHandlerFuncs{
+		AddFunc:    addFunc,
+		UpdateFunc: updateFunc,
+		DeleteFunc: deleteFunc,
 	}
 
-	utillog.Infof("successfully downloaded ConfigMap with UID %q", uid)
-	return payload, "", nil
+	informer := cache.NewSharedInformer(lw, &apiv1.ConfigMap{}, resyncPeriod)
+	informer.AddEventHandler(handler)
```

Comment:
nit: It seems like the interface for `(r *remoteConfigMap) Informer()` could be a bit cleaner, as `addFunc`, `updateFunc`, and `deleteFunc` arent really well-defined in the interface.  Rather than defining them, I would suggest either replacing these inputs with a `cache.ResourceEventHandlerFuncs` parameter, or assuming the caller of `Informer()` will add their desired `ResourceEventHanderFuncs` to the informer returned by `Informer()`.

Reply:
Makes sense, I'll replace them with `cache.ResourceEventHandlerFuncs`.